dask/array/core.py

```diff
@@ -2009,6 +2019,133 @@ def from_array(x, chunks, name=None, lock=False, asarray=True, fancy=True,
     return Array(dsk, name, chunks, dtype=x.dtype)
 
 
+def from_zarr(url, component=None, storage_options=None, **kwargs):
+    """Load array from the zarr storage format
+
+    See https://zarr.readthedocs.io for details about the format.
+
+    Parameters
+    ----------
+    url: str
+        Location of the data. This can include a protocol specifier like s3://
+        for remote data.
+    component: str or None
+        If the location is a zarr group rather than an array, this is the
+        subcomponent that should be loaded, something like ``'foo/bar'``.
+    storage_options: dict
+        Any additional parameters for the storage backend (ignored for local
+        paths)
+    kwargs: passed to zarr's open functions.
+    """
+    import zarr
+    storage_options = storage_options or {}
+    fs, fs_token, path = get_fs_token_paths(
+        url, 'rb', storage_options=storage_options)
+    assert len(path) == 1
+    mapper = get_mapper(fs, path[0])
+    if component is None:
+        z = zarr.open_array(mapper, mode='r', **kwargs)
+    else:
+        z = zarr.open_group(mapper, mode='r', **kwargs)[component]
+    return from_array(z, z.chunks, name='zarr-%s' % url)
+
+
+def to_zarr(arr, url, component=None, storage_options=None,
+            overwrite_group=False, rechunk=False, compute=True,
+            return_stored=False, **kwargs):
+    """Save array to the zarr storage format
+
+    See https://zarr.readthedocs.io for details about the format.
+
+    Parameters
+    ----------
+    arr: dask.array
+        Data to store
+    url: str
+        Location of the data. This can include a protocol specifier like s3://
+        for remote data.
+    component: str or None
+        If the location is a zarr group rather than an array, this is the
+        subcomponent that should be created/over-written.
+    storage_options: dict
+        Any additional parameters for the storage backend (ignored for local
+        paths)
+    overwrite_group: bool
+        If component is not none, this controls whether to create/overwrite the
+        zarr group, or if the group must exits already (but the specific
+        dataset is always overwritten)
+    rechunk: None, False, set of chunks or string
+        Rechunking to be applied to the array before storage, since zarr
+        requires a regular chunks scheme.
+        See ``Array.rechunk``. If False, no rechunk operation is performed,
+        so if the chunks are not regular, an exception is raised.
+    compute, return_stored: see ``store()``
+    kwargs: passed to zarr's open functions, e.g., compression options
+    """
+    import zarr
+    if rechunk is not False:
+        arr = arr.rechunk(rechunk)
```

Comment:
This seems to allow `None` here, which would cause an exception. What would we like to do if `rechunk` is `None`?

Reply:
Right, I haven't implemented any of my suggestions into rechunk yet, but presumably we will need to pick a default. If preferred, this could be called "default".

Reply:
See the example *simple* rechunker I pushed. This could genuinely be a useful one, but as I commented before, I expect there to be a few of these, and we can discuss which makes the best default.

Reply:
Agreed that does seem useful.

Should we pick one for `None` at this stage? Also should we specify `str`s as valid arguments for `rechunk` then?

Reply:
We can pick now, or I can continue to make some more. 
Not sure if that belongs in this PR, though. The set of rechunkers ought to have expensive tests. At this point, I pushed it only so that you could see what I had in mind.

Reply:
Perhaps this should happen in a future PR?  If so we should probably remove these lines.

Reply:
IIUC we decided to add this to avoid locking during writes. If I'm misunderstanding, please feel free to correct me. We can certainly consider locking or other options. `rechunk`ing seems easiest.

That said, there were a lot of ideas floating around about how best to rechunk in different scenarios and make these available to the user. So wouldn't want us to restrict ourselves early while we are still exploring that.

Personally don't have a strong feelings as to whether we keep this in this PR given that last point. We may want to add a note to the docstring about regular chunks being required if we don't supply this. As we already raise for irregular chunks with a nice error message, we should avoid a lot of problems that could come up.

Happy to defer to others on this.